File: lib/domain/model/activity_session_deduplication.dart
import 'dart:math' as math;

import 'activity_models.dart';

const double _duplicateOverlapRatio = 0.85;
const int _duplicateBoundaryToleranceMs = 15 * 60 * 1000;

final DateTime _epoch = DateTime.fromMillisecondsSinceEpoch(0, isUtc: true);

List<ExerciseData> deduplicateExerciseSessions(List<ExerciseData> sessions) {
  if (sessions.length < 2) {
    return [...sessions]..sort((a, b) => b.endTime.compareTo(a.endTime));
  }

  final kept = <ExerciseData>[];
  final sorted = [...sessions]..sort((a, b) {
      final byStart = a.startTime.compareTo(b.startTime);
      if (byStart != 0) return byStart;
      return a.endTime.compareTo(b.endTime);
    });
  for (final session in sorted) {
    final duplicateIndex =
        kept.indexWhere((existing) => _isDuplicateOf(existing, session));
    if (duplicateIndex == -1) {
      kept.add(session);
    } else {
      kept[duplicateIndex] =
          _richerExerciseSession(kept[duplicateIndex], session);
    }
  }

  return kept..sort((a, b) => b.endTime.compareTo(a.endTime));
}

bool _isDuplicateOf(ExerciseData session, ExerciseData other) {
  if (session.exerciseType != other.exerciseType) return false;

  final shorterDuration =
      math.min(math.max(session.durationMs, 0), math.max(other.durationMs, 0));
  if (shorterDuration <= 0) return false;

  final overlapMs = math.min(
        session.endTime.millisecondsSinceEpoch,
        other.endTime.millisecondsSinceEpoch,
      ) -
      math.max(
        session.startTime.millisecondsSinceEpoch,
        other.startTime.millisecondsSinceEpoch,
      );
  if (overlapMs <= 0) return false;

  final startDiff = (session.startTime.millisecondsSinceEpoch -
          other.startTime.millisecondsSinceEpoch)
      .abs();
  final endDiff =
      (session.endTime.millisecondsSinceEpoch - other.endTime.millisecondsSinceEpoch)
          .abs();
  return overlapMs / shorterDuration >= _duplicateOverlapRatio &&
      startDiff <= _duplicateBoundaryToleranceMs &&
      endDiff <= _duplicateBoundaryToleranceMs;
}

ExerciseData _richerExerciseSession(ExerciseData first, ExerciseData second) =>
    _compareRichness(first, second) >= 0 ? first : second;

int _compareRichness(ExerciseData a, ExerciseData b) {
  final byScore = _richnessScore(a).compareTo(_richnessScore(b));
  if (byScore != 0) return byScore;
  final byDuration = a.durationMs.compareTo(b.durationMs);
  if (byDuration != 0) return byDuration;
  return (a.lastModifiedTime ?? _epoch).compareTo(b.lastModifiedTime ?? _epoch);
}

int _richnessScore(ExerciseData session) {
  final metrics = <Object?>[
    session.totalDistanceMeters,
    session.totalCaloriesKcal,
    session.activeCaloriesKcal,
    session.steps,
    session.wheelchairPushes,
    session.averageSpeedMetersPerSecond,
    session.averagePowerWatts,
    session.averageStepsCadenceRate,
    session.averageCyclingCadenceRpm,
    session.averageHeartRateBpm,
    session.floorsClimbed,
    session.elevationGainedMeters,
  ].where((value) => value != null).length;

  return (session.isOpenVitalsEntry ? 1000 : 0) +
      (session.route.status == ExerciseRouteStatus.data ? 200 : 0) +
      math.min<int>(session.route.points.length, 500) +
      metrics * 20 +
      math.min<int>(session.segments.length, 20) * 5 +
      math.min<int>(session.laps.length, 20) * 5 +
      (session.device != null ? 10 : 0) +
      (_isNotBlank(session.title) ? 5 : 0) +
      (_isNotBlank(session.notes) ? 5 : 0);
}

bool _isNotBlank(String? value) => value != null && value.trim().isNotEmpty;

File: app/src/main/kotlin/tech/mmarca/openvitals/domain/model/ActivitySessionDeduplication.kt
package tech.mmarca.openvitals.domain.model

import java.time.Instant
import kotlin.math.max
import kotlin.math.min

private const val DuplicateOverlapRatio = 0.85
private const val DuplicateBoundaryToleranceMs = 15 * 60 * 1000L

fun deduplicateExerciseSessions(sessions: List<ExerciseData>): List<ExerciseData> {
    if (sessions.size < 2) return sessions.sortedByDescending { it.endTime }

    val kept = mutableListOf<ExerciseData>()
    sessions
        .sortedWith(compareBy<ExerciseData> { it.startTime }.thenBy { it.endTime })
        .forEach { session ->
            val duplicateIndex = kept.indexOfFirst { existing -> existing.isDuplicateOf(session) }
            if (duplicateIndex == -1) {
                kept += session
            } else {
                kept[duplicateIndex] = richerExerciseSession(kept[duplicateIndex], session)
            }
        }

    return kept.sortedByDescending { it.endTime }
}

private fun ExerciseData.isDuplicateOf(other: ExerciseData): Boolean {
    if (exerciseType != other.exerciseType) return false

    val shorterDuration = min(durationMs.coerceAtLeast(0L), other.durationMs.coerceAtLeast(0L))
    if (shorterDuration <= 0L) return false

    val overlapMs = min(endTime.toEpochMilli(), other.endTime.toEpochMilli()) -
        max(startTime.toEpochMilli(), other.startTime.toEpochMilli())
    if (overlapMs <= 0L) return false

    val startDiff = (startTime.toEpochMilli() - other.startTime.toEpochMilli()).abs()
    val endDiff = (endTime.toEpochMilli() - other.endTime.toEpochMilli()).abs()
    return overlapMs / shorterDuration.toDouble() >= DuplicateOverlapRatio &&
        startDiff <= DuplicateBoundaryToleranceMs &&
        endDiff <= DuplicateBoundaryToleranceMs
}

private fun richerExerciseSession(first: ExerciseData, second: ExerciseData): ExerciseData =
    compareBy<ExerciseData> { it.richnessScore() }
        .thenBy { it.durationMs }
        .thenBy { it.lastModifiedTime ?: Instant.EPOCH }
        .let { comparator -> if (comparator.compare(first, second) >= 0) first else second }

private fun ExerciseData.richnessScore(): Int =
    (if (isOpenVitalsEntry) 1_000 else 0) +
        (if (route.status == ExerciseRouteStatus.DATA) 200 else 0) +
        route.points.size.coerceAtMost(500) +
        listOfNotNull(
            totalDistanceMeters,
            totalCaloriesKcal,
            activeCaloriesKcal,
            steps,
            wheelchairPushes,
            averageSpeedMetersPerSecond,
            averagePowerWatts,
            averageStepsCadenceRate,
            averageCyclingCadenceRpm,
            averageHeartRateBpm,
            floorsClimbed,
            elevationGainedMeters,
        ).size * 20 +
        segments.size.coerceAtMost(20) * 5 +
        laps.size.coerceAtMost(20) * 5 +
        (if (device != null) 10 else 0) +
        (if (!title.isNullOrBlank()) 5 else 0) +
        (if (!notes.isNullOrBlank()) 5 else 0)

private fun Long.abs(): Long = if (this < 0L) -this else this
